### PR TITLE
[SPARK-17905] [SQL] [TEST] Added test cases for InMemoryRelation

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This pull request adds test cases for the following cases:
- keep all data types with null or without null
- access `CachedBatch` disabling whole stage codegen
- access only some columns in `CachedBatch`

This PR is a part of https://github.com/apache/spark/pull/15219. Here are motivations to add these tests. When https://github.com/apache/spark/pull/15219 is enabled, the first two cases are handled by specialized (generated) code. The third one is a pitfall.

In general, even for now, it would be helpful to increase test coverage.
## How was this patch tested?

added test suites itself

### CODE REVIEW

**Comment by kiszk** [comment]
Jenkins, retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala [hunk 1]
```diff
@@ -20,18 +20,83 @@ package org.apache.spark.sql.execution.columnar
...
 
+  def cachePrimitiveTest(data: DataFrame, dataType: String) {
+    data.createOrReplaceTempView(s"testData$dataType")
+    val storageLevel = MEMORY_ONLY
+    val plan = spark.sessionState.executePlan(data.logicalPlan).sparkPlan
+    val inMemoryRelation = InMemoryRelation(useCompression = true, 5, storageLevel, plan, None)
+
+    assert(inMemoryRelation.cachedColumnBuffers.getStorageLevel == storageLevel)
+    inMemoryRelation.cachedColumnBuffers.collect().head match {
+      case _: CachedBatch => assert(true)
```
Comment (andrewor14) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala]: no need to do assert true here

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala [hunk 2]
```diff
@@ -20,18 +20,83 @@ package org.apache.spark.sql.execution.columnar
...
+      case other => fail(s"Unexpected cached batch type: ${other.getClass.getName}")
+    }
+    checkAnswer(inMemoryRelation, data.collect().toSeq)
+  }
+
+  test("all data type w && w/o nullability") {
+    // all primitives
+    Seq(true, false).map { nullability =>
```
Comment (andrewor14) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala]: can you split these into 2 separate tests? It's more debuggable that way. I would use a private helper method to abstract the logic

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala [hunk 3]
```diff
@@ -58,6 +123,12 @@ class InMemoryColumnarQuerySuite extends QueryTest with SharedSQLContext {
     }.map(Row.fromTuple))
   }
 
+  test("access only some column of the all of columns") {
+    val df = spark.range(1, 100).map(i => (i, (i + 1).toFloat)).toDF("i", "f").cache
+    df.count
```
Comment (andrewor14) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala]: please be explicit with the actions here `count()`, `cache()`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala [hunk 4]
```diff
@@ -246,4 +317,59 @@ class InMemoryColumnarQuerySuite extends QueryTest with SharedSQLContext {
...
+      val schemas = dataTypes.zipWithIndex.map { case (dataType, index) =>
+        StructField(s"col$index", dataType, true)
+      }
+      val rdd = sparkContext.makeRDD(Seq(Row.fromSeq(data)))
+      val df = spark.createDataFrame(rdd, StructType(schemas))
+      val row = df.persist.take(1).apply(0)
+      checkAnswer(df, row)
+    }
+
+    withSQLConf(SQLConf.WHOLESTAGE_MAX_NUM_FIELDS.key -> "2") {
```
Comment (andrewor14) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala]: similarly can you split these into multiple smaller unit tests?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala [hunk 5]
```diff
@@ -20,18 +20,83 @@ package org.apache.spark.sql.execution.columnar
...
 import org.apache.spark.sql.types._
-import org.apache.spark.storage.StorageLevel.MEMORY_ONLY
+import org.apache.spark.storage.StorageLevel._
 
 class InMemoryColumnarQuerySuite extends QueryTest with SharedSQLContext {
   import testImplicits._
 
   setupTestData()
 
+  def cachePrimitiveTest(data: DataFrame, dataType: String) {
```
Comment (andrewor14) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala]: minor: `private def`

**Review by andrewor14 — CHANGES_REQUESTED** [review]
@kiszk This looks good, a few minor code organization comments.

**Comment by andrewor14** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala (hunk 5 above)
Comment (kiszk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala (hunk 1 above)
Comment (kiszk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala]: i see

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala (hunk 3 above)
Comment (kiszk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala]: Done

**Comment by kiszk** [comment]
@andrewor14 thank you for your review. Could you please see it again?

**Comment by andrewor14** [comment]
LGTM, merging into master 2.1 thanks.

**Comment by andrewor14** [comment]
@kiszk is there a JIRA associated specifically with adding tests for `InMemoryRelation`?

**Comment by kiszk** [comment]
Thank you for pointing out an JIRA issue. I made a correction to a JIRA entry.

**Comment by gatorsmile** [comment]
Done. Corrected the JIRA. Thanks!

**Pull request closed without merging**